### PR TITLE
Add configuration file documentation to README

### LINKED ISSUE #344: Document Support for Config Files
# 💡 Summary #
While ScubaGear now supports config files and some example config files have been created, this isn't documented anywhere in the Readme.

## Motivation and context ##
This would make users more aware that this is an option and help them know how to use it (for example, what flag do you pass ScubaGear to specify your config file?).

## Implementation notes ##
Might be good to include some example commands that make use of the example configs already created. Might also be good to let the users know some of the features that can only be achieved by use of a config file, like AAD conditional access policy exceptions.

## Acceptance criteria ##
- [ ] The Readme informs the users that ScubaGear has config file support
- [ ] The Readme informs the users how to specify a config file

### PR DESCRIPTION
**Pull request by schrolla**
#  #

## 🗣 Description ##

README.md file update to provide some sample config files to document typical use cases with
the -ConfigFilePath option

## 💭 Motivation and context ##

Provide information to user to show typical usage of config files and examples
of how they may be combined with overrides from the command line.   This
allows the user to organize and utilize config files in an efficient way for their
particular use without a lot of duplicaiton.

Closes #344 

## 🧪 Testing ##

Linter checks done on sample yaml config files

## ✅ Pre-approval checklist ##

- [X] This PR has an informative and human-readable title.
- [X] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [X] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [ ] Rebase feature branch against parent branch if needed

## ✅ Post-merge checklist ##

- [ ] Check that all tests pass after feature branch merged into parent branch

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -126,8 +138,9 @@ Get-Help -Name Invoke-SCuBA -Full
...
+  
+  If -ConfigFilePath is specified, default values will be provided for parameters that are not explicitly put in the config file.   These default values are provided explicitly in the config file template to help guide the user, but they may be omitted from the file.   Command line values options can now be provided with the -ConfigFilePath. This should reduce the number of config files needed.  Examples might be switching the M365 environment variable from commercial to gcc, switching the tenant used or supplying credentials so they do not need to be resident in the config file (which would facilitate sharing). The file path defaults to the same directory where the script is executed. The file path must point to a valid configuration file. It can be either a relative or absolute path. The file can be used to specify both standard tool parameters as well as custom parameters used by the Azure Active Directory (AAD) product assessment. See [See Scuba-Config Parameter Configuration File Syntax and Examples](#invoke-scuba-parameter-configuration-file-syntax-and-examples) and [AAD Conditional Access Policy Exemptions](#aad-conditional-access-policy-exemptions) for more details.
```
Comment (Sloane4) [inline README.md:142]:
```suggestion
  If a -ConfigFilePath is specified, default values will be provided for parameters that are not explicitly put in the config file. These default values are provided explicitly in the config file template to guide the user, but they can be omitted from the file. Command line values options can now be provided with the -ConfigFilePath. This should reduce the number of config files needed. Examples might be: switching the M365 environment variable from commercial to gcc, switching the tenant used, or supplying credentials so they do not need to be in the config file (to facilitate sharing). The config file path defaults to the same directory where the script is executed. The file path must point to a valid configuration file with a relative or absolute path. The file can be used to specify both standard tool parameters as well as custom parameters used by the Azure Active Directory (AAD) product assessment. See [See Scuba-Config Parameter Configuration File Syntax and Examples](#invoke-scuba-parameter-configuration-file-syntax-and-examples) and [AAD Conditional Access Policy Exemptions](#aad-conditional-access-policy-exemptions) for more details.
```

File: README.md [hunk 2]
```diff
@@ -149,6 +162,105 @@ Get-Help -Name Invoke-SCuBA -Full
 
 - **$OutPath** refers to the folder path where the output JSON and the HTML report will be created. Defaults to the same directory where the script is executed. This parameter is only necessary if an alternate report folder path is desired. The folder will be created if it does not exist.
 
+### Invoke-SCuBA Parameter Configuration File Syntax and Examples
+Most of the `Invoke-SCuBA` parameters supplied on the comment line can be placed into
+a configuration file with the path specified by `-ConfigFilePath` parameter. It should be noted that the following parameters are support only on the command line
```
Comment (Sloane4) [inline README.md:167]:
```suggestion
a configuration file with the path specified by `-ConfigFilePath` parameter. It should be noted that the following parameters are supported only on the command line.
```

File: README.md [hunk 3]
```diff
@@ -149,6 +162,105 @@ Get-Help -Name Invoke-SCuBA -Full
...
+- ConfigFilePath
+- Version
+- DarkMode
+- Quiet
+
+All authentication parameters must be supplied on the command line if a non-interactive login is desired. 
+
+All of the configuration file examples referenced below are in the sample-config-files directory and the examples assume a Invoke-SCuBA is run in that directory.  Each example
+shows the sample config file name and a command line example with it
```
Comment (Sloane4) [inline README.md:176]:
```suggestion
shows the sample config file name and a command line example with it.
```

File: README.md [hunk 4]
```diff
@@ -149,6 +162,105 @@ Get-Help -Name Invoke-SCuBA -Full
...
+Most of the `Invoke-SCuBA` parameters supplied on the comment line can be placed into
+a configuration file with the path specified by `-ConfigFilePath` parameter. It should be noted that the following parameters are support only on the command line
+- ConfigFilePath
+- Version
+- DarkMode
+- Quiet
+
+All authentication parameters must be supplied on the command line if a non-interactive login is desired. 
+
+All of the configuration file examples referenced below are in the sample-config-files directory and the examples assume a Invoke-SCuBA is run in that directory.  Each example
```
Comment (Sloane4) [inline README.md:175]:
```suggestion
All of the configuration file examples referenced below are in the sample-config-files directory and the examples assume a Invoke-SCuBA is run in that directory. Each example
```

File: README.md [hunk 5]
```diff
@@ -149,6 +162,105 @@ Get-Help -Name Invoke-SCuBA -Full
...
+shows the sample config file name and a command line example with it
+
+All authentication parameters shown are samples only.  THe user must supply correct parameters
```
Comment (Sloane4) [inline README.md:178]:
```suggestion
All authentication parameters shown are samples only. The user must supply correct parameters.
```

File: README.md [hunk 6]
```diff
@@ -149,6 +162,105 @@ Get-Help -Name Invoke-SCuBA -Full
...
+
+**Minimal Use** : config file `minimal_config.yaml`
+Minimal use is typically specifying a product name  and an M365 environment variable.  In this example product is entered a a single value
```
Comment (Sloane4) [inline README.md:181]:
```suggestion
Minimal use is typically specifying a product name and an M365 environment variable. In this example product is entered a a single value.
```

File: README.md [hunk 7]
```diff
@@ -149,6 +162,105 @@ Get-Help -Name Invoke-SCuBA -Full
...
+Command line 
+`Invoke-SCuBA -ConfigFilePath minimal_config.yaml`
+
+Command line with override of M365Environment
+```
+Invoke-SCuBA -M365Environment gcc -ConfigFilePath minimal_config.yaml
+```
+
+**Typical Use** : config file `typical_config.yaml`
+Multiple products (can be specified in a list) and an M365 environment variable.  Note that unneeded values are commented out.
```
Comment (Sloane4) [inline README.md:196]:
```suggestion
Multiple products (can be specified in a list) and an M365 environment variable. Note that unneeded values are commented out.
```

File: README.md [hunk 8]
```diff
@@ -149,6 +162,105 @@ Get-Help -Name Invoke-SCuBA -Full
...
+```
+Command line with Auth Parameters
+```
+Invoke-SCuBA -Organization abcdef.organization.com `
+             -AppID 0123456789abcdef01234566789abcde `
+             -CertificateThumbprint: fedcba9876543210fedcba9876543210fedcba98 `
+             -ConfigFilePath typical_config.yaml
+```
+**Credential Use** : config file `creds_config.yaml`
+Credentials can be supplied in the config file.  However the file is not encrypted so
```
Comment (Sloane4) [inline README.md:215]:
```suggestion
Credentials can be supplied in the config file. However the file is not encrypted so
```

File: README.md [hunk 9]
```diff
@@ -149,6 +162,105 @@ Get-Help -Name Invoke-SCuBA -Full
...
+CertificateThumbprint: fedcba9876543210fedcba9876543210fedcba98
+```
+Command line with override of product names
+```
+Invoke-SCuBA -ProductNames  defender -ConfigFilePath typical_config.yaml
+```
+
+**Full Use**: config file `full_config.yaml`
+This example shows all of the parameters supported by ScubaConfig specified in the config
+file.  Any one of these parameters may be commented out.  If not specified ( or commented out ).  ScubaConfig will supply the default value for it unless overridden on the command line.  This default value does not apply to authentication parameters.
```
Comment (Sloane4) [inline README.md:237]:
```suggestion
file. Any one of these parameters may be commented out. If not specified ( or commented out ). ScubaConfig will supply the default value for it unless overridden on the command line. This default value does not apply to authentication parameters.
```

**Review by Sloane4 — CHANGES_REQUESTED** [review]
Runs with out error, however double spacing after a period is not used in our other documents & looks jarring in the markdown. Request removal.

**Review by Sloane4 — APPROVED** [review]

File: README.md [hunk 10]
```diff
@@ -149,6 +162,105 @@ Get-Help -Name Invoke-SCuBA -Full
...
 
+### Invoke-SCuBA Parameter Configuration File Syntax and Examples
+Most of the `Invoke-SCuBA` parameters supplied on the comment line can be placed into
+a configuration file with the path specified by `-ConfigFilePath` parameter. It should be noted that the following parameters are supported only on the command line.
+- ConfigFilePath
+- Version
+- DarkMode
+- Quiet
+
+All authentication parameters must be supplied on the command line if a non-interactive login is desired.
```
[the remark this answers is not in the document]
Comment (kennethpalmer) [reply README.md:173]: No should stand corrected on this.  My wording was not clear.  The distinction is that there is no default for authentication parmeters.  If not in the config file,  any missing ones must be supplied in the command line.  If it is in the config file, the command line one will override.  Let me think through this wording

[the remark this answers is not in the document]
Comment (kennethpalmer) [reply README.md:173]: I'll put the suggested change in a separate commit from the batch ( prefer not to do changes in the gui unless trivial )

**Pull request merged**